NanoProd/plugins/TauSpinnerTableProducer.cc

```diff
@@ -0,0 +1,245 @@
+        name.replace(pos, 1, "minus");
+      out.push_back(std::make_pair(name, val));
+    }
+    return out;
+  }
+
+  void printModuleInfo(edm::ParameterSet const &config) const {
+    std::string header = (boost::format("%-39s%39s") % config.getParameter<std::string>("@module_type") %
+                          config.getParameter<std::string>("@module_label"))
+                             .str();
+    std::cout << std::string(78, '-') << "\n";
+    std::cout << header << "\n";
+    // std::cout << "Produces: " << flow << "\n";
+    std::cout << boost::format("%-15s : %-60s\n") % "Input" % config.getParameter<edm::InputTag>("input").encode();
+    std::cout << boost::format("%-15s : %-60s\n") % "Branch" % config.getParameter<std::string>("branch");
+    std::string thetaStr;
+    for (const auto &theta : theta_vec_)
+      thetaStr += theta.first + ",";
+    std::cout << boost::format("%-15s : %-60s\n") % "Theta" % thetaStr;
+  }
+
+  edm::EDGetTokenT<edm::View<reco::GenParticle>> genPartsToken_;
+  std::string branch_;
+  std::vector<std::pair<std::string, double>> theta_vec_;
+  int bosonPdgId_;
+  std::string tauSpinnerPDF_;
+  bool ipp_;
+  int ipol_;
+  int nonSM2_;
+  int nonSMN_;
+  double cmsE_;
+};
+
+TauSpinnerTableProducer::TauSpinnerTableProducer(const edm::ParameterSet &config)
+    : genPartsToken_(consumes(config.getParameter<edm::InputTag>("input"))),
+      branch_(config.getParameter<std::string>("branch")),
+      bosonPdgId_(25),                                             //Higgs, to be configurable?
+      tauSpinnerPDF_(config.getParameter<std::string>("pdfSet")),  //"NNPDF30_nlo_as_0118"
+      ipp_(true),                                                  //pp collisions
+      ipol_(0),
+      nonSM2_(0),
+      nonSMN_(0),
+      cmsE_(config.getParameter<double>("cmsE"))  //cms energy in GeV, 13000.0
+{
+  theta_vec_ = nameAndValue(config.getParameter<std::vector<double>>("theta"));
+
+  printModuleInfo(config);
+
+  //state that we use tauola/tauspinner resource
+  usesResource(edm::SharedResourceNames::kTauola);
+
+  produces<nanoaod::FlatTable>();
+}
+
+void TauSpinnerTableProducer::getBosons(edm::RefVector<edm::View<reco::GenParticle>> &bosons,
+                                        const edm::View<reco::GenParticle> &parts) const {
+  unsigned idx = 0;
+  for (auto part : parts) {
+    if (std::abs(part.pdgId()) == bosonPdgId_ && part.isLastCopy()) {
+      edm::Ref<edm::View<reco::GenParticle>> partRef(&parts, idx);
+      bosons.push_back(partRef);
+    }
+    ++idx;
+  }
+}
+
+reco::GenParticleRef TauSpinnerTableProducer::getLastCopy(const reco::GenParticleRef &part) const {
+  reco::GenParticleRef last_copy(part);
+  if (part->statusFlags().isLastCopy() && part->statusFlags().fromHardProcess())
+    return last_copy;
+  for (const auto &daughter : part->daughterRefVector()) {
+    if (daughter->pdgId() == part->pdgId()) {
+      last_copy = getLastCopy(daughter);
+    }
+  }
+  return last_copy;
+}
```

Comment:
```suggestion
reco::GenParticleRef TauSpinnerTableProducer::getLastCopy(const reco::GenParticleRef &part) {
  if (part->statusFlags().isLastCopy())
    return part;
  for (const auto &daughter : part->daughterRefVector()) {
    if (daughter->pdgId() == part->pdgId() && part->statusFlags().fromHardProcess()) {
      return getLastCopy(daughter);
    }
  }
  throw std::runtime_error("getLastCopy: no last copy found");
}
```

Reply:
Why `part->statusFlags().fromHardProcess()` in l.135? Should it be either checked in l.132 or replaced by `daughter->statusFlags().fromHardProcess()`?

Reply:
Finally implemented l.135 as follows:
`   if (daughter->pdgId() == part->pdgId() && daughter->statusFlags().fromHardProcess()) {`
Please cross-check.

Reply:
Indeed, it was a copy-paste error on my side. Thanks for fixing it!